src/sampling.jl

```diff
@@ -42,43 +42,44 @@ function direct_sample!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
 end
 direct_sample!(a::AbstractArray, x::AbstractArray) = direct_sample!(Random.GLOBAL_RNG, a, x)
 
-# check whether we can use T to store indices 1:n exactly
-for T in unique!([Int,UInt,Int64,UInt64,Int128,UInt128])
-    @eval isexact(n, ::Type{$T}) = true
-end
-isexact(n, ::Type{T}) where {T<:Integer} = n ≤ typemax(T)
-isexact(n, ::Type{BigInt}) = false # bigint is too expensive
-for T in (Float32, Float64)
-    @eval isexact(n, ::Type{$T}) = n ≤ maxintfloat($T)
-end
-isexact(n, ::Type{Complex{T}}) where {T} = isexact(n, T)
-isexact(n, ::Type{Rational{T}}) where {T} = isexact(n, T)
-isexact(n, T) = false
+# check whether we can use T to store indices 1:n exactly, and
+# use some heuristics to decide whether it is beneficial for k samples
+# (true for a subset of hardware-supported numeric types)
+_storeindices(n, k, ::Type{T}) where {T<:Integer} = n ≤ typemax(T)
+_storeindices(n, k, ::Type{T}) where {T<:Union{Float32,Float64}} = k < 22 && n ≤ maxintfloat(T)
+_storeindices(n, k, ::Type{Complex{T}}) where {T} = _storeindices(n, k, T)
+_storeindices(n, k, ::Type{Rational{T}}) where {T} = k < 16 && _storeindices(n, k, T)
+_storeindices(n, k, T) = false
```

Comment:
Looks like this fallback isn't tested? I guess `BigFloat` would do?

Reply:
`_storeindices` is only called for `Base.HWNumber` types, all of which I currently have methods for; I added this fallback in case `Base.HWNumber` someday adds more types that I haven't handled.

Reply:
I added a test by "manually" calling `_storeindices`.